staking/slash/double-sign.go

```diff
@@ -1,52 +1,251 @@
+}
+
+// Application ..
+type Application struct {
+	TotalSlashed, TotalSnitchReward *big.Int
+}
+
+func (a *Application) String() string {
+	s, _ := json.Marshal(a)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (e Evidence) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Moment
+		ProposalHeader string `json:"header"`
+	}{e.Moment, e.ProposalHeader.String()})
+}
+
+// Records ..
+type Records []Record
+
+// func (r Records) MarshalJSON() ([]byte, error) {
+
+// 	return json.Marshal(struct {
+// 		Slashes []Record `json:""`
+// 	}{r})
+// }
+
+func (r Records) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (r Record) MarshalJSON() ([]byte, error) {
+	reporter, offender :=
+		common2.MustAddressToBech32(r.Reporter),
+		common2.MustAddressToBech32(r.Offender)
+	return json.Marshal(struct {
+		ConflictingBallots
+		Evidence         Evidence `json:"evidence"`
+		Beneficiary      string   `json:"beneficiary"`
+		AddressForBLSKey string   `json:"offender"`
+	}{r.ConflictingBallots, r.Evidence, reporter, offender})
+}
+
+func (e Evidence) String() string {
+	s, _ := json.Marshal(e)
+	return string(s)
+}
+
+func (r Record) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
 
 // Verify checks that the signature is valid
 func Verify(candidate *Record) error {
+	fmt.Println("need to verify the slash", candidate)
+
 	return nil
 }
 
+var (
+	oneHundredPercent = numeric.NewDec(1)
+	fiftyPercent      = numeric.MustNewDecFromStr("0.5")
+)
+
+func validatorSlashApply(
+	wrapper *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	slashTrack *Application,
+) *staking.ValidatorWrapper {
+	// Kick them off forever
+	wrapper.Banned = true
+	// What is the post reduced amount
+	discounted := numeric.NewDecFromBigInt(
+		wrapper.MinSelfDelegation,
+	).Mul(rate).TruncateInt()
+	// The slashed token will be half burned and
+	// half credited to the reporter’s balance.
+	half := numeric.NewDecFromBigInt(
+		new(big.Int).Sub(wrapper.MinSelfDelegation, discounted),
+	).Mul(fiftyPercent).TruncateInt()
+	state.SubBalance(wrapper.Address, half)
+	state.AddBalance(reporter, half)
+	slashTrack.TotalSlashed.Add(slashTrack.TotalSlashed, half)
+	slashTrack.TotalSnitchReward.Add(slashTrack.TotalSnitchReward, half)
+	wrapper.MinSelfDelegation = discounted
+	fmt.Println("after validator slash application", slashTrack.String())
+	return wrapper
+}
+
+func delegatorSlashApply(
+	wrapper *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	slashTrack *Application,
+) *staking.ValidatorWrapper {
+	for _, delegator := range wrapper.Delegations {
+		discounted := numeric.NewDecFromBigInt(
+			delegator.Amount,
+		).Mul(rate).TruncateInt()
+		half := numeric.NewDecFromBigInt(
+			new(big.Int).Sub(delegator.Amount, discounted),
+		).Mul(fiftyPercent).TruncateInt()
+		state.SubBalance(wrapper.Address, half)
```

Comment:
The actual slashing should not be  from the validator's balance, but from the validator's Delegation.Amount and Undelegation.Amount. The calculation of how much token to slash should be based on the staked token at the epoch of double signing. So it should be based on the validator snapshot of that specific epoch when the double sign happened.